WTVCO: Wavetable VCO Hemisphere Applet

beau-seidon opened the pull request:
this PR is for code review, the first commits are the "alpha release".
i am currently implementing the changes discussed in the app-feedback thread.
there will be some more commits shortly!

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+    WaveForms waveform[3];
+    std::vector<int8_t> wt[4];
+
+    bool quantize = true; // unused, delete?
```
Totally just my opinion, but I wouldn't bother with quantization here. General practice in eurorack is to quantize with dedicated quantizers or in sequencers.

beau-seidon replied:
removed. quantization can be done via a different applet or outboard gear. also updated the UI. i updated the pitch calculation to use tideslite.h resources, and now the trigger inputs do pitch range shifting. the full range is now 1 Hz, to 8000 Hz, with some fun aliasing effects at the top end.

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+
+    void gfxRenderWave(int w) {
+        for (size_t x = 0; x < WT_SIZE; x+=4) {
+            int y = 44 - (((wt[w][x] >= 0) * Proportion(wt[w][x], 127, 16)) - (wt[w][x] < 0) * (Proportion(abs(wt[w][x]), 127, 16)));
```
`Proportion` can handle negatives I believe, so this can just be `44 + Proportion(wt[w][x], 127, 16)`.

beau-seidon replied:
updated. i read a comment when i first started that said it only returns positive values, but i pulled it back up and that's just just for ProportionCV override in HemisphereApplet.h. i guess the wrong info stuck with me.

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+        std::vector<int8_t> waveform(size);
+        for (size_t i = 0; i < size; ++i) {
+            double t = static_cast<double>(i) / size;
+            waveform[i] = static_cast<int8_t>(127 * std::sin(TWO_PI * t));
```
I wouldn't use `std::sin` because of code size. You can do `arm_sin_q15` or `arm_sin_q31` instead.

Something like:
```
q15_t t = static_cast<q15_t>(i * 32767 / size);
waveform[i] = arm_sin_q15(t) >> 8;
```

beau-seidon replied:
couldn't get it to work at first, but then i googled it and found arm_math.h and now i have superpowers.

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+
+    std::vector<int8_t> GenerateWaveForm_RandomStep(size_t size) {
+        std::vector<int8_t> waveform(size);
+        std::random_device rd;
+        std::mt19937 gen(rd());
+        std::uniform_int_distribution<int> dist(-128, 127);
```
Just use [random()](https://docs.arduino.cc/language-reference/en/functions/random-numbers/random/)

beau-seidon replied:
im going to mess with this one and all the other AI generated waveforms tomorrow night. i have them commented out for now.

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+        return waveform;
+    }
+
+    std::vector<int8_t> GenerateWaveForm_SinusoidalStep(size_t size) {
+        std::vector<int8_t> waveform(size);
+        for (size_t i = 0; i < size; ++i) {
+            waveform[i] = static_cast<int8_t>(127 * std::round(std::sin(TWO_PI * i / size)));
+        }
+        return waveform;
+    }
```
Isn't this just going to be a square wave?

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+    std::vector<int8_t> GenerateWaveForm_Noise(size_t size) {
+        std::vector<int8_t> waveform(size);
+        for (size_t i = 0; i < size; ++i) {
+            waveform[i] = static_cast<int8_t>(rand() % 256 - 128); // Random int8_t values
```
You can use `random(-128, 128)`

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+        std::vector<int8_t> waveform(size);
+        for (size_t i = 0; i < size; ++i) {
+            double t = static_cast<double>(i) / size;
+            double value = std::exp(-5 * t); // Exponential decay
```
Exponential decay can also just be computed by multiplying the previous value by a fixed fraction each iteration. Then you can avoid the `std:exp` call.

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+        std::vector<int8_t> waveform(size);
+        for (size_t i = 0; i < size; ++i) {
+            double t = static_cast<double>(i) / size;
+            waveform[i] = static_cast<int8_t>(127 * std::cos(TWO_PI * t));
```
There's `arm_cos_*` functions in addition to `arm_sin_*` functions.

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+        for (size_t i = 0; i < size; ++i) {
+            double t = static_cast<double>(i) / size;
+            double sineValue = std::sin(TWO_PI * t);
+            sineValue = std::clamp(sineValue, -clip, clip);
```
For anything that won't just be compiled away, you probably want to prefer arduino functions over std functions. So here, I would use `constrain` over `clamp`.

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+        std::vector<int8_t> waveform(size);
+        for (size_t i = 0; i < size; ++i) {
+            double t = static_cast<double>(i) / size;
+            waveform[i] = static_cast<int8_t>(127 * std::fabs(std::sin(TWO_PI * t)));
```
Just use `abs` after casting. Want to eliminate any std floating point stuff for code size.

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,691 @@
+        for (size_t i = 0; i < size; ++i) {
+            double t = static_cast<double>(i) / size;
+            double sineValue = std::sin(TWO_PI * t);
+            waveform[i] = static_cast<int8_t>(127 * std::max((int)sineValue, 0));
```
There's a `ham_MAX` so you can avoid the std call. Just use it after cast.

qiemem reviewed (commented):
Mostly comments about avoid `std` calls. Check flash usage with just the basic applet without any std calls to get a baseline so that you can tell if anything you do use is suddenly pulling in something big. Didn't repeat comments we discussed in discord, but I can stick them in if you'd like to discuss in context.

beau-seidon commented:
thank you for the comments! don't bother putting the discord stuff in here if you don't want, unless you think it will benefit someone reading it in the future. i'm updating the code currently so some of the previous comments will hopefully be obsolete

qiemem commented on software/src/applets/WTVCO.h:
```diff
@@ -0,0 +1,692 @@
+    }
+
+    void Controller() {
+        if (Clock(0)) pitch_range_shift = constrain(pitch_range_shift - 1, 0, 8);
+        if (Clock(1)) pitch_range_shift = constrain(pitch_range_shift + 1, 0, 8);
+
+        if (Changed(0)) pitch = In(0);
+        phase_inc = ComputePhaseIncrement(pitch);
+        phase += phase_inc;
+        uint8_t phase_acc_msb = (uint8_t)(phase >> (24 - pitch_range_shift));
```
No need to adjust frequency range with bitshifts like this. I recommend making `pitch` an editable parameter that is then added to the incoming CV. This will give you crazy slow LFO to audio rate range for free. And v/oct scaling works well regardless of baseline pitch, so no need to adjust your CV range or anything. That's what most digital LFOs I can think of do.

beau-seba replied:
sweet thank you! i'll experiment with it this evening.

beau-seidon replied:
sweet thank you! i'll experiment with it this evening.

xmacex commented:
Uuh cool. This self-patched via a slew limiter was a very refreshing exploration of harsh noise drone croaking tonight. Thank you for the work :)

I noticed that there is a bug, probably already a known one¹, in the parameters page where changing the modulation targets are not reflected on the screen. This gets confusing when e.g. volume or the divider are pushed out of audio range. Functionally everything appears to be working fine on the synthesis side, it seems to be just the UI.

¹ I am not sure if non-main branch bug reports are welcome?

beau-seidon commented:
i'm glad you like it! it was very fun to create and i learned a lot in the process.

nope didn't know about the UI bug... thanks! i just found it in the code and fixed it.
a recent commit to help optimize the applet size missed an array index variable, so it printed the mod destination label for the first channel on both lines.

im currently in the process of making some more changes, i'll push the update in an hour or two

beau-seidon committed:
WTVCO: params UI bug fix, CV changes, pitch range changes

- CV inputs now offset the base parameter values set by the encoder

- octave shift triggers now directly adjust the pitch parameter

- works much better as an LFO now, low BASE period is 62 seconds (0.01613 Hz)

- associated the Trending icon, and changed applet id to 67

- replaced Cumulus's duplicate id 74 with an unused one

beau-seidon commented:
fixed. CV input now offsets the base values set by the encoders, and pitch is WAY more adjustable. works great as a funky LFO now. other changes noted in commit message
